### DISCUSSION TITLE
Faust .dsp's that don't Compile

### DISCUSSION THREAD

**Discussion by jonrichings** [pschatzmann/arduino-audio-tools#220] [Q&A] [answered]
I've been trying a number of compiled .dsp made in the FaustIDE and #included in a generic .ino wrapper based on streams-generator-faust-i2c, just swapping out the noise.h for the new compiled .dsp.  Here's a list of the Faust examples that I've tried and the errors they create.  It looks to me like most of them have something missing for dealing with MetaGlue and UIGlue, whatever that is.

```
#include "volume.h" // works!
//#include "flute.h" // variable or field 'metadatamydsp' declared void
//#include "VirtualAnalog.h" // doesn't compile: C:\Users\Jon\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioFaust.h:16:5: note:   no known conversion for argument 1 from 'mydsp' to 'dsp&'C:\Users\Jon\Arduino\libraries\arduino-audio-tools\src/AudioLibs/AudioFaust.h:13:7: note: candidate: 'audio_tools::FaustStream::FaustStream(audio_tools::FaustStream&)'class FaustStream : public AudioStreamX {
//#include "osc.h" // doesn't compile: no matching function for call to 'audio_tools::FaustStream::FaustStream(mydsp&, audio_tools::AudioKitStream&)'
//#include "sawtoothLab.h" // C:\Users\Jon\Arduino\flook-1\sawtoothLab.c:101:20: error: unknown type name 'MetaGlue' and C:\Users\Jon\Arduino\flook-1\sawtoothLab.c:377:42: error: unknown type name 'UIGlue'
//#include "noise.h" // doesn't compile: C:\Users\Jon\Arduino\flook-1\noise.c:58:20: error: unknown type name 'MetaGlue' and C:\Users\Jon\Arduino\flook-1\noise.c:112:42: error: unknown type name 'UIGlue'
```
Hope this helps.  I can post complete compiler error logs if required.

**Comment by sletz** [comment]
You are generating pure C code here, is that expected? Why to C++ ?

**Comment by pschatzmann** [comment]
When you export make sure that you select 
Platform: source
Architecture: cplusplus

I suggest to rename the generated .cpp file into .h and import it into the Arduino Sketch.
Compare your noise example with mine: streams-faust-i2s

Please note that e.g. the flute will not fit into the ESP32 RAM and you will need to use PSRAM instead. So far I tried to avoid to use this because I wanted my sketches to run on any ESP32...

**Reply by jonrichings** [reply]
I knew that! ;)  I don't know why I started outputting faust.dsp in c instead of cpp.  My apologies.  Yes, those errors all went away when I used the correct file type.

And, you are correct, many of the .dsp don't compile for lack of internal ram (I'm impressed that the compiler knows that).  Some, like flute are just barely bigger than the space available.  It would be nice if there was an easy way to use that psram, especially for Faust apps that will often be chubby.

Lastly, I have got noise.dsp and volume.dsp to work but some others compile but generate a runtime error:

_[E] AudioFaust.h : 168 - Faust expects input - you need to provide and AudioStream in the constructor_

Which, I assume means the Faust .dsp/.cpp/.h is expecting an input and your code is catching that.  But as far as I can tell the .dsp's that cause this error should need no input.  Do they need a dummy input?  How should I add that to: 
```
mydsp dsp;
FaustStream faust(dsp);
AudioKitStream out; 
StreamCopy copier(out, faust); 
```

**Reply by sletz** [reply]
You can possibly use the [custom memory manager](https://faustdoc.grame.fr/manual/architectures/#custom-memory-manager) model to have the architecture machinery decide what to put on RAM or on PSRAM.

**Reply by jonrichings** [reply]
Stephane, that's getting a bit beyond my skill level but it looks like a way to optimize Faust for multiple blocks of memory.  I think I still need to tell the ESP32 that it should USE multiple blocks.  ESP32 Wrover has about 0.5MB of internal ram and 4/8MB (depending on who you believe) of external PSRAM.  But it has no easy way of using the external.

https://faustdoc.grame.fr/tutorials/esp32/ talks about doing this via the ESP32 IDF but not through the Arduino IDE.  It also mentions:

_Note: We're currently modifying the Faust compiler to automatize this process when a specific option is given to it._  

But I'm guessing that hasn't happened yet, or at least, not for the Faust Web IDE.

I've looked at other forums about implementing external PSRAM on the ESP32 but it is not clear to me how this would work within Phil's framework.

**Reply by sletz** [reply]
Assuming external PSRAM can be activated on ESP32, the [custom memory manager](https://faustdoc.grame.fr/manual/architectures/#custom-memory-manager) is the new way to do things on Faust side (I mean the https://faustdoc.grame.fr/tutorials/esp32/ is now a bit old and we should try to use this custom-memory-manager model instead).

**Reply by pschatzmann** [reply]
I agree, to use PSRAM properly we need generated code that uses a memory manager.  I was looking in the FAUST IDE for a way to specify the -mem option but did not find anything.  Using Faust LIVE with the -mem options did not provide the expected result either. What did I miss ?

**Comment by pschatzmann** [comment]
If you check in my code you get this message when if (p_dsp->getNumInputs()>0)

**Reply by jonrichings** [reply]
This error occurred with sawtoohLab.dsp.  I did not realise that this .dsp had an input option.  My mistake.

**Comment by sletz** [comment]
Working here with the latest FL release:
<img width="881" alt="Capture d’écran 2022-05-28 à 11 51 53" src="https://user-images.githubusercontent.com/1430487/170820277-5400555c-7694-4c10-97ac-68eff0f9c1d3.png">

**Comment by pschatzmann** [comment]
Hmm, I upgraded, but it still does not work:

- Faust Live About gives Dynamic Faust Compiler version 2.40.5

Additional Compilation Step - Faust Compiler Option is -mem

As a result I get a cpp file with the follwing info in the header
version: "1.0"
Compilation options: -lang cpp -es 1 -mcd 16 -single -ftz 0

And no memory manager...

**Reply by sletz** [reply]
Same Faust Compiler version 2.40.5 here, tested on macOS.  Are you on linux?

**Reply by pschatzmann** [reply]
No, an old osx version.
What compilation web service url do you have in your Preferences ? 

About Faust Live displays down at the cancel button the following information: Version 2.5.10 by GRAME...

**Reply by sletz** [reply]
Should be OK. The "Additional Compilation Step" is actually done inside the application, no interaction with the remote compiler. Do other options like `-vec -lv 32` or `-mcd 4` work?

**Reply by pschatzmann** [reply]
I decided to restart the machine. Now I am not able to open any examples any more and I am getting an error -mem can only be used with cpp or occp backends. I think it is not possible to set the backend w/o opening an example.

Do you know how to get out of this?

**Comment by sletz** [comment]
They are two zones in the UI:
- "FAUST Compiler options" : the one use for te "LLVM backend" (dynamic compilation
- Additional Compilation Step : the one where you can generate other file (like .cpp, possibly SVG...) 

The `-mem` one should be tested in  - Additional Compilation Step.

Otherwise you can possibly try the binary release here: https://github.com/grame-cncm/faust/releases

**Comment by pschatzmann** [comment]
I rebooted my machine and now I am locked out. I am getting the following error when I want to open an example:
![Screen Shot 2022-05-28 at 13 33 16](https://user-images.githubusercontent.com/2078916/170823732-74d4ccf1-5828-4088-9ef9-66e0d8707463.png)
And the example does not open. To my knowledge I can only change this option in an example. So I am locked out.

Do you know how to get out of this ?

**Reply by sletz** [reply]
Arrrgh !!  ((-; probably some saved preferences to remove... but where ? trying to remember that...

**Reply by sletz** [reply]
Doing `rm -rf .FaustLive-CurrentSession-2.4` in your home folder should solve the issue.

**Reply by sletz** [reply]
But again using an installed Faust binary version should be easier...

**Reply by pschatzmann** [reply]
The rm was helping, but in spite of the restart still no memory manager. 

I can try to binary version if you tell me how to execute the option Source Code CPP on the command line.

**Reply by pschatzmann** [reply]
faust -mem fluteStk.dsp -o flute.h
looks much better!

**Comment by jonrichings** [comment]
Wow, glad you two worked that out and not me!

So, I'm assuming the Faust memory manager is now working via Faust Live,
but it probably isn't added to the Web IDE yet?

And Phil, I guess this still needs changes your libs to take advantage of
it?

**Comment by pschatzmann** [comment]
I just committed my changes. 

Please note that the API has slightly changed:
- [Example w/o PSRAM / noise](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit/streams-faust_noise-i2s)
- [Example with PSRAM and Memory Manager / flute](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-audiokit/streams-faust_flute-i2s)

**Comment by sletz** [comment]
Do they use the memory manager?

**Comment by pschatzmann** [comment]
The example with SPMEM is using the memory manager. 
The one w/o has been compiled w/o -mem - so it's old style...

**Comment by sletz** [comment]
With [dsp_memory_manager](https://github.com/pschatzmann/arduino-audio-tools/commit/f5af9dd943bc56ec2feecdb8fb925faf19fe2ff5#diff-cfabf70ca2f7daa0f43ce4f1fbc28cad5a1e1c6b866ae3752a3a46f2076ee817R150) would is make sense to keep small zones and high I/O in ram and only allocate bigger ones in PSRAM?

**Comment by pschatzmann** [comment]
That would have been my understanding. In my latest commit am using the following logic:
        void* result = is_psram && size > PSRAM_LIMIT ? ps_malloc(size) : malloc(size);
I am currently using PSRAM_LIMIT as 1K

**Comment by jonrichings** [comment]
Phil, Stephane, will this work with .dsp created in the faustide, or is this limited to options -mem compiled in faust live?

**Comment by pschatzmann** [comment]
I was not able to generate this in FaustLive and only managed to do it via the command line: The command is documented in the comment of the sketch

**Comment by jonrichings** [comment]
@pschatzmann and @sletz 

Faust's latest memory manager and Phil's new support, along with the noise.h and fluteMidi.h, that Phil managed to compile, work perfectly.

Stephane, do you think the problems Phil had with making the memory manager work in FaustLive can be easily fixed?  And what are the chances of this option be added to the FaustIDE (which is the only way I'm smart enough to compile .dsp).

**Comment by sletz** [comment]
@jonrichings  you should try, again working here: 
- in the windows parameter, use the Additional Compilation Steps", "Faust Compiler Options" and add something like `-mem -o foo.cpp -O /path/to/folder` to create the foo.cpp C++ file using the memory_manager
<img width="881" alt="170820277-5400555c-7694-4c10-97ac-68eff0f9c1d3" src="https://user-images.githubusercontent.com/1430487/171260801-b9f30ea9-5d66-4727-9490-292a23164585.png">

**Reply by jonrichings** [reply]
I've never been able to get FaustLive to run properly on a windows setup.  It often crashes and when it does run it runs VERY slowly, taking several seconds to respond to any dsp controls, and the sound is always slowed way down.  I have to close it and start from clean boot before opening a new file or it crashes.  It's pretty much unusable on my PC (old but i5 and 16GB ram, Win10).  I wish I could.

**Comment by pschatzmann** [comment, marked as answer]
I suggest to use the command line compiler: see https://github.com/pschatzmann/arduino-audio-tools/wiki/Faust

**Reply by jonrichings** [reply]
I can't get the command line compiler to install on Windows.  I get _"Warning! PATH too long installer unable to modify PATH!"_.  I've mentioned this elsewhere but havn't found a solution.

**Comment by pschatzmann** [comment]
Are you sure that this is not just a warning and that you can set the path after the installation?

**Reply by jonrichings** [reply]
The installer quits at that point.

**Reply by jonrichings** [reply]
I just tried again and realised there was an option to not add the path.  That completed installation.  I guess I have use absolute paths to Faust in the cmd line now?

**Comment by pschatzmann** [comment]
You can try to update the path in the Control Panel

**Reply by jonrichings** [reply]
Got it.  Took me a while to find the correct way to write the path.  Now working.  Thanks Phil :)

**Comment by jonrichings** [comment]
I am considering redesigning my hardware to use a esp32-pico-mini, but it only has 2MB of PSRAM.  Your changes to use PSRAM have opened a big door to Faust but I am unsure as to how much PSRAM I really need.  Is there a simple way to read or measure the amount that any given Faust + Arduino app will need?
